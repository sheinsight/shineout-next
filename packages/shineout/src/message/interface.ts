export type { MessageProps as Props } from './message.type';
