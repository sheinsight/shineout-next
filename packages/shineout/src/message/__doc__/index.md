---
name: Message
group: Feedback
---

# Title

Message 消息提示
Message

# Describe

用户进行操作后，系统对该操作行为内容作出的反馈
Display global messages as feedback in response to user operations.

# Example
