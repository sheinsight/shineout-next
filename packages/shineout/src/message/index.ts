import Message from './message';

type RefMessage = typeof Message;

export interface MessageComponent extends RefMessage {
  displayName: string;
}

const MessageComp: MessageComponent = Message as MessageComponent;

MessageComp.displayName = 'ShineoutMessage';

export default MessageComp;
