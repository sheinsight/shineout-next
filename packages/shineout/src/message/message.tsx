import { Message, MessageFuncArg, MessageOptions as MessageOptionsUnStyle } from '@sheinx/base';
import { useAlertStyle, useMessageStyle } from '@sheinx/shineout-style';

export type MessageOptions = Omit<MessageOptionsUnStyle, 'jssStyle'>;
const createMessage = (
  type: 'show' | 'success' | 'info' | 'warn' | 'warning' | 'danger' | 'error',
) => {
  return (
    content: MessageFuncArg['content'],
    duration?: MessageFuncArg['duration'],
    options: MessageOptions = {},
  ) =>
    Message[type](content, duration, {
      ...options,
      jssStyle: {
        message: useMessageStyle,
        alert: useAlertStyle,
      },
    });
};
const MessageWithStyle = {
  ...Message,
  show: createMessage('show'),
  success: createMessage('success'),
  info: createMessage('info'),
  warn: createMessage('warn'),
  warning: createMessage('warning'),
  danger: createMessage('danger'),
  error: createMessage('error'),
};

export default MessageWithStyle;
