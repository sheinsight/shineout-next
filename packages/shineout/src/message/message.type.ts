import { MessageOptions as UnStyledMessageProps } from '@sheinx/base';

export type MessageOptions = Omit<UnStyledMessageProps, 'jssStyle'>;
