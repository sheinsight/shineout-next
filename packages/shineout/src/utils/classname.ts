import classnames from 'classnames';
import { config } from '@sheinx/base';

interface CSSModule {
  [key: string]: string;
}

/**
 * create a new className generate function, add namespace, handle css module
 * @param style - object; for css module
 * @param module - string
 * @param prefix - string, default value is 'shineout'
 * * */
export default (style: CSSModule, module: string, prefix: string = config.prefix) =>
  (...args: any) => {
    const className = classnames(...args);
    if (!className) return '';

    const ns = `${prefix}${module ? `-${module}` : '-'}`;
    let list: (string | number)[] = className
      .split(' ')
      .map((c) => (c === '_' ? ns : `${ns}-${c}`));
    if ((config as any).cssModule) {
      list = list.map((c: any) => style[c] || c);
    }
    return list.join(' ');
  };
