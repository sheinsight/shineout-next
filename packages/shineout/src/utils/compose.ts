export function compose(...funcs: any) {
  if (funcs.length === 0) {
    return (arg: any) => arg;
  }
  const last = funcs[funcs.length - 1];
  const rest = funcs.slice(0, -1);
  return (...args: typeof last) =>
    rest.reduceRight((composed: any, f: any) => f(composed), last(...args));
}
