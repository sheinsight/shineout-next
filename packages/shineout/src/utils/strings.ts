import { util } from "@sheinx/hooks";

const { devUseWarning } = util;

export function capitalize(str: string) {
  if (typeof str !== 'string') {
    devUseWarning.error('str should be a string');
  }
  return str && str[0].toUpperCase() + str.slice(1);
}

export function substitute(str: string | Function, obj: { [x: string]: any }) {
  if (typeof str === 'string') {
    if (str.indexOf('{') < 0) {
      return str;
    }

    return str.replace(/\\?\{([^{}]+)\}/g, (match, name) => {
      if (match.charAt(0) === '\\') {
        return match.slice(1);
      }
      return obj[name] === null || obj[name] === undefined ? '' : obj[name];
    });
  }
  if (typeof str === 'function') {
    let val = str(obj);
    if (val === obj && typeof val === 'object') {
      val = Object.assign({}, obj);
    }
    return val;
  }

  return '';
}

export function removeProtocol(url: string) {
  if (url.indexOf('http') !== 0) return url;
  try {
    const { href, protocol } = new URL(url);
    return href.slice(protocol.length);
  } catch (error) {
    return url;
  }
}

export function getRTLPosition(position: string) {
  if (!position) return position;
  // position.replace('left', 'right').replace('right', 'left')
  if (position.indexOf('left') !== -1) {
    return position.replace('left', 'right');
  }
  if (position.indexOf('right') !== -1) {
    return position.replace('right', 'left');
  }
  return position;
}

export function getDirectionIconName(mode = 'left', double = false) {
  const rtl = true;
  if (mode === 'left') {
    if (rtl) {
      return double ? 'ArrowRightDouble' : 'AngleRight';
    }
    return double ? 'ArrowLeftDouble' : 'AngleLeft';
  }

  if (rtl) {
    return double ? 'ArrowLeftDouble' : 'AngleLeft';
  }
  return double ? 'ArrowRightDouble' : 'AngleRight';
}
