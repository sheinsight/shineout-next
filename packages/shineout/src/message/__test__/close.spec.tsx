import { render, cleanup, fireEvent, waitFor, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MessageTest } from './message.spec'
import { delay, createClassName, textContentTest, styleTest, attributesTest } from '../../tests/utils';
import MessageClose from '../__example__/04-close';
import MessageManualClose from '../__example__/05-manual-close';

const SO_PREFIX = 'message';
const originClasses = ['wrapper', 'message'];
const originItemClasses = [''];
const { wrapper, message } = createClassName(
  SO_PREFIX,
  originClasses,
  originItemClasses,
);

const SO_PREFIX_Alert = 'alert';
const originAlertClasses = ['close', 'title'];
const originAlertItemClasses = [''];
const {
  close: closeClassName,
  title: titleClassName,
} = createClassName(SO_PREFIX_Alert, originAlertClasses, originAlertItemClasses);

beforeAll(() => {
  jest.useFakeTimers();
});
afterAll(() => {
  jest.runAllTimers();
  cleanup()
});

afterEach(cleanup);

describe('Message[Close]', () => {
  afterEach(cleanup);
  test('should render when set onClose', async () => {
    const closeFn = jest.fn();
    const { container, rerender } = render(<MessageTest options={{ onClose: closeFn }} />);
    fireEvent.click(container.querySelector('button')!);
    await waitFor(async () => {
      await delay(200);
      fireEvent.click(document.querySelector(closeClassName)!);
    });
    await waitFor(async () => {
      await delay(200);
      expect(closeFn.mock.calls.length).toBe(1);
    });
    rerender(<MessageClose />);
    fireEvent.click(container.querySelector('button')!);
    await waitFor(async () => {
      await delay(200);
      const temp = document.querySelector(closeClassName)!
      fireEvent.click(temp.querySelector('svg')!);
    });
    // expect(document.querySelector(wrapper)).not.toBeInTheDocument();
  });

  test('should render when set title', async () => {
    const { container } = render(<MessageTest options={{ title: 'hello' }} />);
    fireEvent.click(container.querySelector('button')!);
    await waitFor(async () => {
      await delay(200);
      textContentTest(document.querySelector(titleClassName)!, 'hello');
      fireEvent.click(document.querySelector(closeClassName)!);
    });
    await waitFor(async () => {
      await delay(200);
      // expect(document.querySelector(wrapper)).not.toBeInTheDocument();
    });
  });

  test('should render manual close', async () => {
    const { container } = render(<MessageManualClose />);
    fireEvent.click(container.querySelector('button')!);
    await waitFor(async () => {
      await delay(200);
    });
    expect(document.querySelector(wrapper)).toBeInTheDocument();
    // fireEvent.click(document.querySelector('a')!);
    // await waitFor(async () => {
    //   await delay(200);
    // });
    // expect(document.querySelector(wrapper)).not.toBeInTheDocument();
  });
})
describe('Message[Options]', () => {
  test('should render when set top', async () => {
    const { container } = render(<MessageTest options={{ top: '-20px' }} />);
    fireEvent.click(container.querySelector('button')!);
    // await waitFor(async () => {
    //   await delay(200);
    //   styleTest(document.querySelector(message)!, 'top: -20px;');
    //   fireEvent.click(document.querySelector(closeClassName)!);
    // });
    // await waitFor(async () => {
    //   await delay(200);
    //   expect(document.querySelector(wrapper)).not.toBeInTheDocument();
    // });
  });
  // test('should render when set position is middle', async () => {
  //   const { container } = render(<MessageTest options={{ position: 'middle' }} />);
  //   fireEvent.click(container.querySelector('button')!);
  //   screen.debug()
  //   await waitFor(() => {attributesTest(document.querySelector(wrapper)!, 'data-soui-position', 'middle')})
  // })
})