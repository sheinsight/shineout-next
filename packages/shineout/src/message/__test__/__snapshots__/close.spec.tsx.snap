// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message[Base] should render correctly  1`] = `
<div>
  <button
    class="c-pzb818 soui-button-secondary c-3br65q soui-button-button"
    dir="ltr"
    type="button"
  >
    <span>
      Show
    </span>
  </button>
  <button
    class="ctqq9eq soui-button-primary c-3br65q soui-button-button"
    dir="ltr"
    style="margin-inline-start: 24px;"
    type="button"
  >
    <span>
      Info
    </span>
  </button>
  <button
    class="c42guub soui-button-success c-3br65q soui-button-button"
    dir="ltr"
    style="margin-inline-start: 24px;"
    type="button"
  >
    <span>
      Success
    </span>
  </button>
  <button
    class="c-hhihj8 soui-button-warning c-3br65q soui-button-button"
    dir="ltr"
    style="margin-inline-start: 24px;"
    type="button"
  >
    <span>
      Warn
    </span>
  </button>
  <button
    class="c-2orx19 soui-button-danger c-3br65q soui-button-button"
    dir="ltr"
    style="margin-inline-start: 24px;"
    type="button"
  >
    <span>
      Error
    </span>
  </button>
  <button
    class="c-pzb818 soui-button-secondary c-3br65q soui-button-button"
    dir="ltr"
    style="margin-inline-start: 24px;"
    type="button"
  >
    <span>
      Close All
    </span>
  </button>
</div>
`;

exports[`Message[Base] should render correctly  2`] = `
<div>
  <button
    class="c-pzb818 soui-button-secondary c-3br65q soui-button-button"
    dir="ltr"
    type="button"
  >
    <span>
      Duration 10 s.
    </span>
  </button>
  <button
    class="c-pzb818 soui-button-secondary c-3br65q soui-button-button"
    dir="ltr"
    style="margin-inline-start: 24px;"
    type="button"
  >
    <span>
      Manually close
    </span>
  </button>
</div>
`;

exports[`Message[Base] should render correctly  3`] = `
<div>
  <div
    style="display: flex; align-items: center; margin-bottom: 24px;"
  >
    <div
      class="c2yfuoi soui-radio-group"
    >
      <div
        class="clnvxwm soui-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="chxtlw1 soui-radio-indicator-wrapper"
        >
          <span
            class="c-4u63tq soui-radio-indicator"
            tabindex="1"
          />
        </span>
        <span
          class="c6yu2se soui-radio-desc"
        >
          top
        </span>
      </div>
      <div
        class="clnvxwm soui-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="chxtlw1 soui-radio-indicator-wrapper"
        >
          <span
            class="c-4u63tq soui-radio-indicator"
            tabindex="1"
          />
        </span>
        <span
          class="c6yu2se soui-radio-desc"
        >
          middle
        </span>
      </div>
      <div
        class="clnvxwm soui-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="chxtlw1 soui-radio-indicator-wrapper"
        >
          <span
            class="c-4u63tq soui-radio-indicator"
            tabindex="1"
          />
        </span>
        <span
          class="c6yu2se soui-radio-desc"
        >
          top-left
        </span>
      </div>
      <div
        class="clnvxwm soui-radio-wrapper c-863kv soui-radio-wrapper-checked"
      >
        <input
          checked=""
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="chxtlw1 soui-radio-indicator-wrapper"
        >
          <span
            class="c-4u63tq soui-radio-indicator"
            tabindex="1"
          />
        </span>
        <span
          class="c6yu2se soui-radio-desc"
        >
          top-right
        </span>
      </div>
      <div
        class="clnvxwm soui-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="chxtlw1 soui-radio-indicator-wrapper"
        >
          <span
            class="c-4u63tq soui-radio-indicator"
            tabindex="1"
          />
        </span>
        <span
          class="c6yu2se soui-radio-desc"
        >
          bottom-left
        </span>
      </div>
      <div
        class="clnvxwm soui-radio-wrapper"
      >
        <input
          style="display: none;"
          tabindex="0"
          type="radio"
        />
        <span
          class="chxtlw1 soui-radio-indicator-wrapper"
        >
          <span
            class="c-4u63tq soui-radio-indicator"
            tabindex="1"
          />
        </span>
        <span
          class="c6yu2se soui-radio-desc"
        >
          bottom-right
        </span>
      </div>
    </div>
  </div>
  <button
    class="c-pzb818 soui-button-secondary c-3br65q soui-button-button"
    dir="ltr"
    type="button"
  >
    <span>
      Show message.
    </span>
  </button>
</div>
`;

exports[`Message[Base] should render default 1`] = `
<div
  class="c6jfqe2 soui-message-wrapper"
  data-soui-position="top"
>
  <div
    class="c-ra3u50 soui-message-item c-rj5kxj soui-message-item-show "
  >
    <div
      class="c-41y6gi soui-message-message c8pzyb2 soui-alert-alert cy5b50u soui-alert-warning"
      data-soui-type="message"
    >
      <div
        class="cvacykr soui-alert-content"
      >
        <div
          class="c-200wmd soui-alert-text"
        >
          Test
        </div>
      </div>
      <div
        class="c8r3qx6 soui-alert-close"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M4.6129 3.2097L4.70711 3.29289L12 10.585L19.2929 3.29289C19.6834 2.90237 20.3166 2.90237 20.7071 3.29289C21.0676 3.65338 21.0953 4.22061 20.7903 4.6129L20.7071 4.70711L13.415 12L20.7071 19.2929C21.0976 19.6834 21.0976 20.3166 20.7071 20.7071C20.3466 21.0676 19.7794 21.0953 19.3871 20.7903L19.2929 20.7071L12 13.415L4.70711 20.7071C4.31658 21.0976 3.68342 21.0976 3.29289 20.7071C2.93241 20.3466 2.90468 19.7794 3.2097 19.3871L3.29289 19.2929L10.585 12L3.29289 4.70711C2.90237 4.31658 2.90237 3.68342 3.29289 3.29289C3.65338 2.93241 4.22061 2.90468 4.6129 3.2097Z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="c-ra3u50 soui-message-item c-rj5kxj soui-message-item-show "
  >
    <div
      class="c-41y6gi soui-message-message c8pzyb2 soui-alert-alert cy5b50u soui-alert-warning"
      data-soui-type="message"
    >
      <div
        class="cvacykr soui-alert-content"
      >
        <div
          class="c-200wmd soui-alert-text"
        >
          Test
        </div>
      </div>
      <div
        class="c8r3qx6 soui-alert-close"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M4.6129 3.2097L4.70711 3.29289L12 10.585L19.2929 3.29289C19.6834 2.90237 20.3166 2.90237 20.7071 3.29289C21.0676 3.65338 21.0953 4.22061 20.7903 4.6129L20.7071 4.70711L13.415 12L20.7071 19.2929C21.0976 19.6834 21.0976 20.3166 20.7071 20.7071C20.3466 21.0676 19.7794 21.0953 19.3871 20.7903L19.2929 20.7071L12 13.415L4.70711 20.7071C4.31658 21.0976 3.68342 21.0976 3.29289 20.7071C2.93241 20.3466 2.90468 19.7794 3.2097 19.3871L3.29289 19.2929L10.585 12L3.29289 4.70711C2.90237 4.31658 2.90237 3.68342 3.29289 3.29289C3.65338 2.93241 4.22061 2.90468 4.6129 3.2097Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`Message[Base] should render default 2`] = `
<div
  class="c6jfqe2 soui-message-wrapper"
  data-soui-position="top"
>
  <div
    class="c-ra3u50 soui-message-item c-rj5kxj soui-message-item-show "
  >
    <div
      class="c-41y6gi soui-message-message c8pzyb2 soui-alert-alert cy5b50u soui-alert-warning"
      data-soui-type="message"
    >
      <div
        class="cvacykr soui-alert-content"
      >
        <div
          class="c-200wmd soui-alert-text"
        >
          Test
        </div>
      </div>
      <div
        class="c8r3qx6 soui-alert-close"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M4.6129 3.2097L4.70711 3.29289L12 10.585L19.2929 3.29289C19.6834 2.90237 20.3166 2.90237 20.7071 3.29289C21.0676 3.65338 21.0953 4.22061 20.7903 4.6129L20.7071 4.70711L13.415 12L20.7071 19.2929C21.0976 19.6834 21.0976 20.3166 20.7071 20.7071C20.3466 21.0676 19.7794 21.0953 19.3871 20.7903L19.2929 20.7071L12 13.415L4.70711 20.7071C4.31658 21.0976 3.68342 21.0976 3.29289 20.7071C2.93241 20.3466 2.90468 19.7794 3.2097 19.3871L3.29289 19.2929L10.585 12L3.29289 4.70711C2.90237 4.31658 2.90237 3.68342 3.29289 3.29289C3.65338 2.93241 4.22061 2.90468 4.6129 3.2097Z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="c-ra3u50 soui-message-item c-rj5kxj soui-message-item-show "
  >
    <div
      class="c-41y6gi soui-message-message c8pzyb2 soui-alert-alert cy5b50u soui-alert-warning"
      data-soui-type="message"
    >
      <div
        class="cvacykr soui-alert-content"
      >
        <div
          class="c-200wmd soui-alert-text"
        >
          Test
        </div>
      </div>
      <div
        class="c8r3qx6 soui-alert-close"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M4.6129 3.2097L4.70711 3.29289L12 10.585L19.2929 3.29289C19.6834 2.90237 20.3166 2.90237 20.7071 3.29289C21.0676 3.65338 21.0953 4.22061 20.7903 4.6129L20.7071 4.70711L13.415 12L20.7071 19.2929C21.0976 19.6834 21.0976 20.3166 20.7071 20.7071C20.3466 21.0676 19.7794 21.0953 19.3871 20.7903L19.2929 20.7071L12 13.415L4.70711 20.7071C4.31658 21.0976 3.68342 21.0976 3.29289 20.7071C2.93241 20.3466 2.90468 19.7794 3.2097 19.3871L3.29289 19.2929L10.585 12L3.29289 4.70711C2.90237 4.31658 2.90237 3.68342 3.29289 3.29289C3.65338 2.93241 4.22061 2.90468 4.6129 3.2097Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
